### ISSUE TITLE
Diagnostic, Maintenance and Sleep Modes for Bots

### ISSUE THREAD

**Issue by mokun** [mars-sim/mars-sim#1161]
**Is your feature request related to a problem? Please describe.**
- mars-sim used to put robots to sleep when not being in active use
- Currently, the default state is Charging when not in use by other tasks

**Describe the rationale**
- When not in use, robots should be able to run internal diagnostic to uncover issues ahead of time
- As an equipment, Robots need to be well-maintained
- Engineers and another person on that robot's specialty should be given a task to refine the programming of a robot

**Describe the goal(s) you have in mind**
- Create and differentiate Diagnostic, Maintenance and Sleep mode and define what they do 
- Put robots to either one of these modes when not being in active use
- Create a programming task for engineers

**Specifications**
- pre 3.7.0 build 9025

**Additional context**
- When in maintenance mode, meet the engineer in the lab or the workshop
- Define a closet location to stow a bot when not in use for a longer period

**Comment by mokun** [comment]
@bevans2000 

Can we go back to making *Sleep mode* possible for robots ? 

When there's nothing to do, the robot should NOT be in the charging state.

Charging would stop at 95%. Then it should go to the *Power Save* mode.

**Comment by mokun** [comment]
@bevans2000 

I'll add a PowerSave task for bots to go into *Power Save* mode to conserve power.

It will be called in BotTaskManager just like currently with getChargeTaskMap() with the base rating score of 1.

```java
	private static synchronized TaskCache getChargeTaskMap() {
		if (chargeMap == null) {
			chargeMap = new TaskCache("Robot Charge", null);
			TaskJob chargeJob = new AbstractTaskJob("Charge", new RatingScore(1D)) {
				
				private static final long serialVersionUID = 1L;

				@Override
				public Task createTask(Robot robot) {
					return new Charge(robot, null);
				}	
			};
			chargeMap.put(chargeJob);
		}
		return chargeMap;
	}
```

**Comment by bevans2000** [comment]
The Sleep task is very specific to Person so I would want a new PowerSave Task dedicated to Robots. I like that idea, you would move them out of a charging station so it is free for other Robots to charge?

**Comment by mokun** [comment]
@bevans2000 

Can we make robot an equipment type again ?

I'm putting together a manufacturing process for making bots.

It'll need to become an equipment, just like an EVA suit or a vehicle that can be assembled.

**Comment by bevans2000** [comment]
No need, follow the pattern we use for manufacturing Vehicles. Changing the hierarchy of Robots is likely to break lots of things

**Comment by mokun** [comment]
hmmm. 

I guess the process outputs an EVA Suit has it as a equipment. 

But the process outputs an explorer rover has it as a vehicle. 


```xml
	<process name="Assemble explorer rover" tech="3" skill="4" work-time="2000" process-time="1000"	tooling="lifting"
			power-required="4.0">
		<inputs>
			<resource name="acetylene" amount="2.0" />		
			<resource name="polystyrene" amount="15" />			
			<part name="aerogel tile" number="80" />					
			<part name="rover control panel" number="1" />			
			<part name="vehicle frame large" number="3" />
			<part name="vehicle chassis panel" number="6" />
...
		<outputs>
			<vehicle name="explorer rover" number="1" />
		</outputs>
	</process>

	<process name="Assemble EVA suit" tech="2" skill="3" work-time="500"
			power-required="0.5">
		<description>
...
		</description>
		<inputs>
			<part name="EVA helmet" number="1" />
			<part name="helmet visor" number="1" />
			<part name="pressure suit" number="1" />
			<part name="coveralls" number="1" />
			<part name="Liquid Cooling Garment" number="1" />
...
		</inputs>
		<outputs>
			<equipment name="EVA suit" number="1" />
		</outputs>
	</process>

```

So logically speaking, the process that outputs a bot will deliver it as a robot.

**Comment by mokun** [comment]
> Don't convert Robot back to Equipment, that would be wrong

Agree. No need. I didn't do that in https://github.com/mars-sim/mars-sim/commit/86a81d3d1ffc3d606ea79c4be63589779afe80a4.

**Comment by mokun** [comment]
I wanna add a new tooling called "Wave Soldering" for assembling various electrical parts together to form another part or equipment.

It's for fabricating an electronic component or any manu process involving a logic board, a circuit, etc.

I know metalwork would always include the need to solder but it's beyond just working with metal, you know.

I asked the Brave AI and it came back with the followings : 

```
Wave soldering is a bulk soldering process used in printed circuit board manufacturing.

Wave soldering is commonly used in mass production due to its efficiency and ability to handle large volumes of components.

The process involves several stages, including fluxing, preheating, and soldering, with the solder wave height and contact time carefully controlled to ensure quality
```

What do you think ? 

```xml
<!-- List of all manufacturing processes. -->
<manufacturing-processes>
	<tooling>
		<tool name="Furnace" description="Equipment used to bake or melt materials" />
		<tool name="3D Printer" description="Generic 3D printer" />
		<tool name="Lifting" description="Lifting equipment for working on large items" />
		<tool name="Metalwork" description="Tools to shape and construct metal items" />
		<tool name="Chemical" description="Ability to perform chemical reactions and transformations" />
	</tooling>
```

So I'll add

```xml
<tool name="Wave Soldering" description="A mass production process of joining metal surfaces together using filler metals" />
```

**Comment by mokun** [comment]
Hand soldering would be more like repairing one thing at time.

For a small scale production, people probably use SMD soldering. Only a small SMD plate ~$15 US is needed.
1. https://www.youtube.com/shorts/1aypeCFuIBg
2. https://www.youtube.com/shorts/xALzJ93Do3c

For large scale production, they use wave soldering but a big machine is needed.
1. https://www.youtube.com/shorts/UZ5d5UYx2jc
2. https://www.youtube.com/shorts/Vc-lf3ulpnA

**Comment by mokun** [comment]
@bevans2000 @Urwumpe 

Robots are subject to wear and tear and are serviceable but we never have robot maintenance until now. 

I'm almost done implementing the basic. Stay tune.
